packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,61 +3,499 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
 
-/// A leaf node in the focus tree that can receive focus.
+import 'binding.dart';
+import 'focus_scope.dart';
+import 'framework.dart';
+
+/// Signature of a callback used by [Focus.onKey] and [FocusScope.onKey]
+/// to receive key events.
 ///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
+/// The [node] is the node that received the event.
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
+
+/// An attachment point for a [FocusNode].
 ///
-/// To request focus, find the [FocusScopeNode] for the current [BuildContext]
-/// and call the [FocusScopeNode.requestFocus] method:
+/// Once created, a [FocusNode] must be attached to the widget tree by its
+/// _host_ [StatefulWidget] via a [FocusAttachment] object. [FocusAttachment]s
+/// are owned by the [StatefulWidget] that hosts a [FocusNode] or
+/// [FocusScopeNode]. There can be multiple [FocusAttachment]s for each
+/// [FocusNode], but the node will only ever be attached to one of them at a
+/// time.
 ///
-/// ```dart
-/// FocusScope.of(context).requestFocus(focusNode);
+/// This attachment is created by calling [FocusNode.attach], usually from the
+/// host widget's [State.initState] method. If the widget is updated to have a
+/// different focus node, then the new node needs to be attached in
+/// [State.didUpdateWidget], after calling [detach] on the previous
+/// [FocusAttachment]. Once detached, the attachment is defunct and will no
+/// longer make changes to the [FocusNode] through [reparent].
+///
+/// Without these attachment points, it would be possible for a focus node to
+/// simultaneously be attached to more than one part of the widget tree during
+/// the build stage.
+class FocusAttachment {
+  /// A private constructor, because [FocusAttachment]s are only to be created
+  /// by [FocusNode.attach].
+  FocusAttachment._(this._node) : assert(_node != null);
+
+  // The focus node that this attachment manages an attachment for. The node not
+  // yet have a parent, or may have been detached from this attachment, so don't
+  // count on this node being in a usable state.
+  final FocusNode _node;
+
+  /// Returns true if the associated node is attached to this attachment.
+  ///
+  /// It is possible to be attached to the widget tree, but not be placed in
+  /// the focus tree (i.e. to not have a parent yet in the focus tree).
+  bool get isAttached => _node._attachment == this;
+
+  /// Detaches the [_node] this attachment point is associated with from the
```

Comment:
The public doc probably shouldn't reference a private variable?

Reply:
Yeah, that was IntelliJ helping me when I took the variable private.  Fixed.